xarray/core/parallel.py

```diff
@@ -0,0 +1,333 @@
+        function will be run eagerly.
+    args: tuple
+        Passed on to func after unpacking. xarray objects, if any, will not be split by
+        chunks.
+    kwargs: dict
+        Passed on to func after unpacking. xarray objects, if any, will not be split by
+        chunks.
+
+    Returns
+    -------
+    A single DataArray or Dataset
+
+    Notes
+    -----
+
+    This function is designed to work with dask-backed xarray objects. See apply_ufunc
+    for a similar function that works with numpy arrays.
+
+    See Also
+    --------
+    dask.array.map_blocks, xarray.apply_ufunc, xarray.Dataset.map_blocks,
+    xarray.DataArray.map_blocks
+    """
+
+    def _wrapper(func, obj, to_array, args, kwargs):
+        if to_array:
+            obj = dataset_to_dataarray(obj)
+
+        result = func(obj, *args, **kwargs)
+
+        for name, index in result.indexes.items():
+            if name in obj.indexes:
+                if len(index) != len(obj.indexes[name]):
+                    raise ValueError(
+                        "Length of the %r dimension has changed. This is not allowed."
+                        % name
+                    )
+
+        return make_dict(result)
+
+    if not isinstance(args, Sequence):
+        raise TypeError("args must be a sequence (for example, a list or tuple).")
+
+    if kwargs is None:
+        kwargs = {}
+
+    elif not isinstance(kwargs, Mapping):
+        raise TypeError("kwargs must be a mapping (for example, a dict)")
+
+    if not dask.is_dask_collection(obj):
+        return func(obj, *args, **kwargs)
+
+    if isinstance(obj, DataArray):
+        # only using _to_temp_dataset would break
+        # func = lambda x: x.to_dataset()
+        # since that relies on preserving name.
+        if obj.name is None:
+            dataset = obj._to_temp_dataset()
+        else:
+            dataset = obj.to_dataset()
+        input_is_array = True
+    else:
+        dataset = obj
+        input_is_array = False
+
+    input_chunks = dataset.chunks
+
+    template = infer_template(
+        func, obj, *args, **kwargs
+    )  # type: Union[DataArray, Dataset]
+    if isinstance(template, DataArray):
+        result_is_array = True
+        template_name = template.name
+        template = template._to_temp_dataset()
+    elif isinstance(template, Dataset):
+        result_is_array = False
+    else:
+        raise TypeError(
+            "func output must be DataArray or Dataset; got %s" % type(template)
+        )
+
+    template_indexes = set(template.indexes)
+    dataset_indexes = set(dataset.indexes)
+    preserved_indexes = template_indexes.intersection(dataset_indexes)
+    new_indexes = set(template_indexes) - set(dataset_indexes)
+    indexes = {dim: dataset.indexes[dim] for dim in preserved_indexes}
+    indexes.update({k: template.indexes[k] for k in new_indexes})
+
+    graph = {}  # type: Dict[Any, Any]
+    gname = "%s-%s" % (dask.utils.funcname(func), dask.base.tokenize(dataset))
```

Comment:
This should maybe depend both on the dataset, and on all inputs to this function.  For example the following results should have different names.:

```python
ds.map_blocks(func, kwargs={"foo": 1})
ds.map_blocks(func, kwargs={"foo": 2})
```

Reply:
Also, has xarray implemented dask.base.normalize_token for DataSets and DataArrays?

Reply:
> This should maybe depend both on the dataset, and on all inputs to this function.

Done.


> Also, has xarray implemented dask.base.normalize_token for DataSets and DataArrays?

Doesn't look like it. (cc @shoyer , @crusaderky )

Reply:
Nope.
```python
>>> a = xarray.DataArray(da.ones(2))
>>> normalize_token(a)
'6743063447a54e87a8bd15ef07733093'
>>> normalize_token(a)
'b2e634be86fb49f6ab7f5deac862f4fe'
```
This looks like a showstopper - I'm afraid you'll have to implement it...

Reply:
In order not to make this PR balloon even more, it may be a good idea to create a new PR for normalize_token, review & merge it, and then go back to work on this one.

Reply:
It's fine to not implement `normalize_token`.  It is safe to not have it implemented, it just means that you won't get deduplication of equivalent computations.  This is slightly inefficient, but doesn't produce wrong results.  I would recommend merging in this PR, and raising an issue for `normalize_token` some other time.

Reply:
Opened #3378